src/workerd/util/checked-queue.h

```diff
@@ -0,0 +1,167 @@
+// a safe-to-use queue that avoids the pitfalls of using std::list directly
+// (such as accidentally dangling references when the list is empty but someone
+// calls front(), etc).
+template <Movable T>
+class Queue final {
+ public:
+  Queue() = default;
+  Queue(Queue<T>&&) = default;
+  Queue<T>& operator=(Queue<T>&&) = default;
+  KJ_DISALLOW_COPY(Queue);
+
+  inline void push(T&& value) {
+    inner.push_back(kj::mv(value));
+  }
+
+  template <typename... Args>
+  T& emplace(Args&&... args) KJ_LIFETIMEBOUND {
+    return inner.emplace_back(std::forward<Args>(args)...);
+  }
+
+  // Pops the front element from the queue, moving it out.
+  // Returns kj::none if the queue is empty.
+  inline kj::Maybe<T> pop() {
+    if (inner.empty()) {
+      return kj::none;
+    }
+    T value = kj::mv(inner.front());
+    inner.pop_front();
+    return kj::mv(value);
+  }
+
+  // Returns a reference to the front element without removing it.
+  // Returns kj::none if the queue is empty.
+  inline kj::Maybe<T&> peek() KJ_LIFETIMEBOUND {
+    if (inner.empty()) {
+      return kj::none;
+    }
+    return inner.front();
+  }
+
+  // Returns a reference to the front element without removing it.
+  // Returns kj::none if the queue is empty.
+  inline kj::Maybe<const T&> peek() const KJ_LIFETIMEBOUND {
+    if (inner.empty()) {
+      return kj::none;
+    }
+    return inner.front();
+  }
+
+  // Returns a reference to the last element without removing it.
+  // Returns kj::none if the queue is empty.
+  inline kj::Maybe<T&> peekBack() KJ_LIFETIMEBOUND {
+    if (inner.empty()) {
+      return kj::none;
+    }
+    return inner.back();
+  }
+
+  // Returns a reference to the last element without removing it.
+  // Returns kj::none if the queue is empty.
+  inline kj::Maybe<const T&> peekBack() const KJ_LIFETIMEBOUND {
+    if (inner.empty()) {
+      return kj::none;
+    }
+    return inner.back();
+  }
+
+  // Drains the queue, moving each element to the callback one at a time.
+  // Returns the number of elements moved.
+  inline size_t drainTo(auto callback) {
+    size_t count = 0;
+    while (!inner.empty()) {
+      callback(KJ_ASSERT_NONNULL(pop()));
+      count++;
+    }
+    return count;
+  }
+
+  // Removes elements from the queue that satisfy the given condition.
+  // Returns the number of elements removed.
+  inline size_t deleteIf(auto callback) {
+    size_t count = 0;
+    auto it = inner.begin();
+    while (it != inner.end()) {
+      if (callback(*it)) {
+        it = inner.erase(it);
+        count++;
+      } else {
+        ++it;
+      }
+    }
+    return count;
+  }
+
+  // Takes the first element in the queue that satisfies the given condition, if any.
+  inline kj::Maybe<T> takeIf(auto callback) {
+    for (auto it = inner.begin(); it != inner.end(); ++it) {
+      if (callback(*it)) {
+        T value = kj::mv(*it);
+        inner.erase(it);
+        return kj::mv(value);
```

Comment:
Generally, you shouldn't need to write `return kj::mv(value);` because it's redundant and it interferes with NRVO (named return value optimization). 

But if there is a reason for it, it might be worth adding a comment to stop people like me from refactoring it in the future.

Reply:
I've added a comment. Specifically, it's not strictly necessary but I wanted it to be clear that the value is being moved and not copied. If it's bothersome then it's perfectly fine to remove it.